pkg/upstream/cluster/loadbalancer.go

```diff
@@ -144,5 +147,146 @@ func (lb *roundRobinLoadBalancer) HostNum(metadata api.MetadataMatchCriteria) in
 	return len(lb.hosts.Hosts())
 }
 
+const default_choice = 2
+
+// leastActiveRequestLoadBalancer choose the host with the least active request
+type leastActiveRequestLoadBalancer struct {
+	*EdfLoadBalancer
+	choice uint32
+}
+
+func newleastActiveRequestLoadBalancer(info types.ClusterInfo, hosts types.HostSet) types.LoadBalancer {
+	lb := &leastActiveRequestLoadBalancer{}
+	if info != nil && info.LbConfig() != nil {
+		lb.choice = info.LbConfig().(*v2.LeastRequestLbConfig).ChoiceCount
+	} else {
+		lb.choice = default_choice
+	}
+	lb.EdfLoadBalancer = newEdfLoadBalancerLoadBalancer(hosts, lb.unweightChooseHost, lb.hostWeight)
+	return lb
+}
+
+func (lb *leastActiveRequestLoadBalancer) hostWeight(host types.Host) float64 {
+	return float64(host.Weight()) / float64((host.HostStats().UpstreamRequestActive.Count() + 1))
+}
+
+func (lb *leastActiveRequestLoadBalancer) unweightChooseHost(context types.LoadBalancerContext) types.Host {
+
+	allHosts := lb.hosts.Hosts()
+	total := len(allHosts)
+	lb.mutex.Lock()
+	defer lb.mutex.Unlock()
+	var candicate types.Host
+	// Choose `choice` times and return the best one
+	// See The Power of Two Random Choices: A Survey of Techniques and Results
+	//  http://www.eecs.harvard.edu/~michaelm/postscripts/handbook2001.pdf
+	for cur := 0; cur < int(lb.choice); cur++ {
+
+		randIdx := lb.rand.Intn(total)
+		tempHost := allHosts[randIdx]
+		if candicate == nil {
+			candicate = tempHost
+			continue
+		}
+		if candicate.HostStats().UpstreamRequestActive.Count() > tempHost.HostStats().UpstreamRequestActive.Count() {
+			candicate = tempHost
+		}
+	}
+	return candicate
+
+}
+
 // TODO:
 // WRR
+
+type EdfLoadBalancer struct {
+	scheduler *edfSchduler
+	hosts     types.HostSet
+	rand      *rand.Rand
+	mutex     sync.Mutex
+	// the method to choose host when all host
+	unweightChooseHostFunc func(types.LoadBalancerContext) types.Host
+	hostWeightFunc         func(host types.Host) float64
+}
+
+func (lb *EdfLoadBalancer) ChooseHost(context types.LoadBalancerContext) types.Host {
+
+	var candicate types.Host
+	targetHosts := lb.hosts.Hosts()
+	total := len(targetHosts)
+	if total == 0 {
+		// Return nil directly if allHosts is nil or size is 0
+		return nil
+	}
+	if total == 1 {
+		// Return directly if there is only one host
+		return targetHosts[0]
+	}
+	for i := 0; i < total; i++ {
+		if lb.scheduler != nil {
+			// do weight selection
+			host := lb.scheduler.Next().(types.Host)
+			lb.scheduler.Add(host, lb.hostWeightFunc(host))
```

Comment:
When call `lb.scheduler.Add` function,  the value of  `edf.currentTime `  may already be another item.

Reply:
ok, i have mv this op to the lock block.